Adding FreeBSD build support.

LadySerenaKitty opened the pull request:
Making all changes required to build on FreeBSD.  Some CMake entries for Linux may have changed, but their functionality remains unchanged.

TheMostDiligent commented on BuildTools/FormatValidation/validate_format_freebsd_implementation.sh:
```diff
@@ -0,0 +1,52 @@
+#   --exclude ../../Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions_inc.h \
+#   --exclude ../../Graphics/GraphicsEngineVulkan/shaders/GenerateMipsCS_inc.h \
+#   --exclude ../../Tests/DiligentCoreAPITest/assets/*
+#
```
This is just a verbatim copy of `validate_format_linux_implementation.sh`, so why is it needed?

LadySerenaKitty replied:
We could probably get away with changing the shebang line to env bash, then it would work cleanly on both.

TheMostDiligent replied:
There is no bin/bash on FreeBSD?

LadySerenaKitty replied:
FreeBSD's base only includes sh, tcsh, and csh.  All other shells must be installed either from Ports (source build) or Packages (binary download).  In FreeBSD, everything that isn't part of base goes in `/usr/local`.  We have a very clear separation of base and installed software.

TheMostDiligent commented on BuildTools/FormatValidation/validate_format_freebsd.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/local/bin/bash
+
+source validate_format_freebsd_implementation.sh
+
+validate_format ../../Common ../../Graphics ../../Platforms ../../Primitives ../../Tests \
+  --exclude ../../Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions.h \
+  --exclude ../../Graphics/HLSL2GLSLConverterLib/include/GLSLDefinitions_inc.h \
+  --exclude ../../Graphics/GraphicsEngineVulkan/shaders \
+  --exclude ../../Graphics/GraphicsEngine.NET \
+  --exclude ../../Tests/DiligentCoreAPITest/assets
+
```
Why is this verbatim copy needed?

TheMostDiligent commented on BuildTools/CMake/BuildUtils.cmake:
```diff
@@ -397,6 +397,8 @@ function(add_format_validation_target MODULE_NAME MODULE_ROOT_PATH IDE_FOLDER)
     elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
         set(RUN_VALIDATION_SCRIPT ./validate_format_linux.sh)
+    elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "FreeBSD")
+        set(RUN_VALIDATION_SCRIPT ./validate_format_freebsd.sh)
```
Why not use `./validate_format_linux.sh`?

TheMostDiligent commented on CMakeLists.txt:
```diff
@@ -197,7 +202,14 @@ elseif(PLATFORM_LINUX)
+    if(PLATFORM_FREEBSD)
+        target_compile_definitions(Diligent-PublicBuildSettings INTERFACE PLATFORM_LINUX=1 PLATFORM_FREEBSD=1)
+        if(EXISTS /usr/local/include)
+            include_directories(/usr/local/include)
```
Diligent's build system does not use global CMake include directories.
All directories are set for targets.

Which include files from `/usr/local/include` are need to be used?

LadySerenaKitty replied:
All the X, Wayland, GL, and Vulcan stuffs, since our base is pretty small.  The vast majority of everything is installed manually.

TheMostDiligent replied:
Diligent uses its own headers for Vulkan. For GL, does `find_package` not provide all the required libraries?
We need to find another solution. Heaving global `include_directories(/usr/local/include)` here and in multiple other places is not a good option.

LadySerenaKitty replied:
CMake's `find_package` finds the libraries just fine (needed for linker), but it won't provide headers.  If CMake provides a way to specify a search location to the compiler for specific headers, I don't know about it, and I would use that instead.

TheMostDiligent commented on Platforms/Linux/src/LinuxPlatformMisc.cpp:
```diff
@@ -28,6 +28,10 @@
 #include <pthread.h>
 
+#ifdef PLATFORM_FREEBSD
+#    include <pthread_np.h>
```
Why is this needed?

LadySerenaKitty replied:
None of the pthread_*_np stuffs is available without this.

TheMostDiligent replied:
Where is this stuff is used?

LadySerenaKitty replied:
On lines 45 and 61.

TheMostDiligent commented on ThirdParty/DirectXShaderCompiler/dxc/WinAdapter.h:
```diff
@@ -342,7 +342,9 @@
 #define __vectorcall
 #define __thiscall
+#ifndef PLATFORM_FREEBSD
 #define __fastcall
```
Is `__fastcall` a recognized keyword when compiling on FreeBSD?

LadySerenaKitty replied:
On this one, the compiler was complaining about fastcall being redefined, as FreeBSD's headers already define fastcall.

TheMostDiligent replied:
Which compiler are you using?

LadySerenaKitty replied:
Believe it or not, the entirety of DiligentEngine was built using the compiler provided by the base system: clang 16.0.6.
![image](https://github.com/DiligentGraphics/DiligentCore/assets/965280/f88cabed-e057-4dc2-999b-015ab3a6743d)

LadySerenaKitty replied:
This is what happens without skipping `__fastcall`:

```
[ 35%] Building CXX object DiligentCore/Graphics/ShaderTools/CMakeFiles/Diligent-ShaderTools.dir/src/DXCompiler.cpp.o
In file included from /usr/home/jlhawkwell/Projects/DiligentEngine/DiligentCore/Graphics/ShaderTools/src/DXCompiler.cpp:42:
In file included from /usr/home/jlhawkwell/Projects/DiligentEngine/DiligentCore/Graphics/ShaderTools/include/DXCompilerBaseLinux.hpp:31:
In file included from /usr/home/jlhawkwell/Projects/DiligentEngine/DiligentCore/Graphics/ShaderTools/../../ThirdParty/DirectXShaderCompiler/dxc/dxcapi.h:37:
/usr/home/jlhawkwell/Projects/DiligentEngine/DiligentCore/Graphics/ShaderTools/../../ThirdParty/DirectXShaderCompiler/dxc/WinAdapter.h:345:9: error: '__fastcall' macro redefined [-Werror,-Wmacro-redefined]
#define __fastcall
        ^
/usr/include/sys/cdefs.h:365:9: note: previous definition is here
#define __fastcall      __attribute__((__fastcall__))
        ^
1 error generated.
gmake[2]: *** [DiligentCore/Graphics/ShaderTools/CMakeFiles/Diligent-ShaderTools.dir/build.make:122: DiligentCore/Graphics/ShaderTools/CMakeFiles/Diligent-ShaderTools.dir/src/DXCompiler.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:3521: DiligentCore/Graphics/ShaderTools/CMakeFiles/Diligent-ShaderTools.dir/all] Error 2
gmake: *** [Makefile:156: all] Error 2
```

TheMostDiligent replied:
I updated DXC headers, so this should work now - please try the latest version

TheMostDiligent replied:
A better way to fix this is to check

`#ifndef __fastcall`

Though I do really not understand why it only complains about `__fastcall`. All other macros are defined in the same way.

TheMostDiligent replied:
What I suggested is to use the `#ifndef __fastcall` guard instead of `#ifndef PLATFORM_FREEBSD`.
Also note that this file is a third-party header so with the next update, it may break again.

TheMostDiligent commented on CMakeLists.txt:
```diff
@@ -125,6 +126,10 @@ else()
         message("Target platform: Linux " ${ARCH})
+    elseif(${CMAKE_SYSTEM_NAME} STREQUAL "FreeBSD")
+        set(PLATFORM_LINUX TRUE CACHE INTERNAL "Target platform: Linux")
+        set(PLATFORM_FREEBSD TRUE CACHE INTERNAL "Target platform: FreeBSD")
```
Defining both `PLATFORM_LINUX` and `PLATFORM_FREEBSD` is a bit inconsistent with the rest of platform variables as only a single variable is expected to be defined.
Will it build if you try to build at as Linux?

LadySerenaKitty replied:
This was simpler for me to do as the vast majority of the Linux stuffs in the build system translates pretty clearly to FreeBSD.  I don't mind rejiggering everything so only PLATFORM_FREEBSD is used, if you want me to do that.

TheMostDiligent replied:
> I don't mind rejiggering everything so only PLATFORM_FREEBSD is used, if you want me to do that.

This will require quite a lot of changes, which are not really needed.
Heaving `PLATFORM_FREEBSD` and `PLATFORM_LINUX` at the same time may be OK.

LadySerenaKitty replied:
It should be fine, as PLATFORM_FREEBSD is only added to the environment if it's running on FreeBSD.  Functionally, all the Linux stuff is untouched.

TheMostDiligent commented on BuildTools/CMakeLists.txt:
```diff
@@ -8,6 +8,8 @@ elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Linux")
 elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
     set(CLANG_FORMAT_EXECUTABLE "${CMAKE_CURRENT_SOURCE_DIR}/FormatValidation/clang-format_mac_10.0.0" CACHE INTERNAL "clang-format executable path")
+elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "FreeBSD")
+    set(CLANG_FORMAT_EXECUTABLE "/usr/local/llvm10/bin/clang-format" CACHE INTERNAL "clang-format executable path")
```
What happens if there is no llvm10 installed? There will be an error that is hard to understand.

I am thinking about disabling format validation completely - it is intended for development, but is not required for build.

LadySerenaKitty replied:
This part here is specifically for the format validation.  This part here (along with the scripts I added) may be omitted entirely if desired.

TheMostDiligent commented on Graphics/Archiver/CMakeLists.txt:
```diff
@@ -132,7 +132,14 @@ if(GL_SUPPORTED OR GLES_SUPPORTED)
     elseif(PLATFORM_LINUX)
-        target_link_libraries(Diligent-Archiver-static PRIVATE GLEW::glew GL X11)
+        if(PLATFORM_FREEBSD)
+            find_package(OpenGL REQUIRED)
```
In particular, when you do this, doesn't OpenGL export all the required lib directories?

LadySerenaKitty replied:
From my expurrience, using the short `-lpthread` form to link to `/usr/lib/libpthread.so` works, but that's also a base system component.  I've rarely seen it work for user-installed stuffs.

TheMostDiligent reviewed (changes requested):
I updated core module, so that format validation will be disabled on platforms that don't have the script.
Can you update to the latest version and try it?

LadySerenaKitty commented:
> I updated core module, so that format validation will be disabled on platforms that don't have the script. Can you update to the latest version and try it?

Format validation is successfully skipped (tested after deleting the scripts I added).  Would you like a new push with the scripts removed?

TheMostDiligent commented:
> Would you like a new push with the scripts removed?

You can force-push to the same branch - the PR will be updated.

LadySerenaKitty commented on BuildTools/CMake/BuildUtils.cmake:
```diff
@@ -399,20 +399,22 @@ function(add_format_validation_target MODULE_NAME MODULE_ROOT_PATH IDE_FOLDER)
-    )
+    if (RUN_VALIDATION_SCRIPT)
+        # Run the format validation script
+        add_custom_command(TARGET ${MODULE_NAME}-ValidateFormatting
+            COMMAND ${RUN_VALIDATION_SCRIPT}
+            WORKING_DIRECTORY "${MODULE_ROOT_PATH}/BuildTools/FormatValidation"
+            COMMENT "Validating ${MODULE_NAME} module's source code formatting..."
+            VERBATIM
+        )
 
-    if(TARGET ${MODULE_NAME}-ValidateFormatting)
-        set_target_properties(${MODULE_NAME}-ValidateFormatting PROPERTIES FOLDER ${IDE_FOLDER})
+        if(TARGET ${MODULE_NAME}-ValidateFormatting)
+            set_target_properties(${MODULE_NAME}-ValidateFormatting PROPERTIES FOLDER ${IDE_FOLDER})
+        endif()
+    else()
+		message(DEBUG "${MODULE_NAME}-ValidateFormatting target will be disabled because format validation script is not available on ${CMAKE_HOST_SYSTEM_NAME} host platform.")
```
I literally have no idea how to do that or fix the mess I caused. :crying_cat_face:

LadySerenaKitty commented on BuildTools/CMakeLists.txt:
```diff
@@ -10,6 +10,3 @@ elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL "Darwin")
 
-if (NOT EXISTS ${CLANG_FORMAT_EXECUTABLE})
-    message(FATAL_ERROR "clang-format executable is not found.")
-endif()
```
No longer complains about this on non-validating systems.

TheMostDiligent commented on Graphics/Archiver/CMakeLists.txt:
```diff
@@ -132,7 +132,14 @@ if(GL_SUPPORTED OR GLES_SUPPORTED)
         target_link_libraries(Diligent-Archiver-static PRIVATE GLEW::glew opengl32.lib)
     elseif(PLATFORM_LINUX)
-        target_link_libraries(Diligent-Archiver-static PRIVATE GLEW::glew GL X11)
+        if(PLATFORM_FREEBSD)
+            find_package(OpenGL REQUIRED)
+            find_package(X11 REQUIRED)
+        else()
+            set(OPENGL_gl_LIBRARY GL)
+            set(X11_LIBRARIES X11)
+        endif()
+        target_link_libraries(Diligent-Archiver-static PRIVATE GLEW::glew ${OPENGL_gl_LIBRARY} ${X11_LIBRARIES})
```
This should no longer be needed - please try getting the latest version

TheMostDiligent commented on Graphics/GraphicsEngineOpenGL/CMakeLists.txt:
```diff
@@ -198,7 +198,14 @@ if(PLATFORM_WIN32)
     set(PRIVATE_DEPENDENCIES ${PRIVATE_DEPENDENCIES} GLESv3 EGL)
 elseif(PLATFORM_LINUX)
-    set(PRIVATE_DEPENDENCIES ${PRIVATE_DEPENDENCIES} GLEW::glew GL X11)
+    if(PLATFORM_FREEBSD)
+        find_package(OpenGL REQUIRED)
+        find_package(X11 REQUIRED)
+    else()
+        set(OPENGL_gl_LIBRARY GL)
+        set(X11_LIBRARIES X11)
+    endif()
+    set(PRIVATE_DEPENDENCIES ${PRIVATE_DEPENDENCIES} GLEW::glew ${OPENGL_gl_LIBRARY} ${X11_LIBRARIES})
```
This should no longer be needed

TheMostDiligent commented on CMakeLists.txt:
```diff
@@ -513,6 +522,9 @@ else()
 endif()
 
+if(PLATFORM_FREEBSD AND EXISTS /usr/local/include)
+	target_include_directories(Diligent-BuildSettings SYSTEM AFTER INTERFACE /usr/local/include)
```
What errors do you get without this include?
This change does not seem right

LadySerenaKitty replied:
It is quite correct.  Without this, anything that isn't part of base will not be found.

TheMostDiligent replied:
What base?

LadySerenaKitty replied:
FreeBSD!  "base" is short for "base operating system" and is the stuff that's made by the FreeBSD Project - kernel and world.  This includes everything outside of `/usr/home` and `/usr/local`, which isn't much.

A mirror of the source repo is available here, so you can see what all is part of FreeBSD's base: https://github.com/freebsd/freebsd-src

TheMostDiligent commented on CMakeLists.txt:
```diff
@@ -520,6 +528,10 @@ else()
 endif()
 
+if(PLATFORM_FREEBSD AND EXISTS /usr/local/include)
+    target_include_directories(Diligent-BuildSettings SYSTEM AFTER INTERFACE /usr/local/include)
```
I tried to find some info about this problem, and I think that the right way to fix this is to set `CMAKE_REQUIRED_INCLUDES` at your project's CMake level:

https://github.com/dotnet/runtime/issues/4073

TheMostDiligent replied:
Did you try setting the `CMAKE_REQUIRED_INCLUDES` in your project CMake?

TheMostDiligent reviewed (changes requested):
Sorry for a long delay with response.
